File: prototype/cogit/refs.py
"""Refs, HEAD, and reflogs: lockfile updates with old-target checks.

Protocol (ADR-0010, closes OQ-006): create <path>.lock exclusively, re-check
expected old value under the lock, write + fsync + atomic rename, then append
the reflog line. A reflog append failure after a successful ref move is
surfaced, not rolled back.
"""

import os
import re

from .errors import ConcurrentUpdateError, CorruptionError, UserError
from .objects import is_oid

REF_SEGMENT_RE = re.compile(r"^[a-z0-9._-]+$")
REFLOG_LINE_RE = re.compile(
    r"^(?P<old>\S+) (?P<new>\S+) (?P<ts>\S+) (?P<actor>\S+) (?P<op>[^:]+): (?P<reason>.*)$"
)


def validate_ref_name(name: str):
    """Validate a full ref name like 'refs/heads/main' per repository-layout-v1."""
    if not name or name.startswith("/") or name.endswith("/"):
        raise UserError(f"refs: invalid ref name '{name}'")
    if name.endswith(".lock") or "@{" in name or "\\" in name or ".." in name:
        raise UserError(f"refs: invalid ref name '{name}'")
    for segment in name.split("/"):
        if not segment or not REF_SEGMENT_RE.match(segment):
            raise UserError(f"refs: invalid ref segment '{segment}' in '{name}'")


class RefStore:
    def __init__(self, cogit_dir):
        self.cogit_dir = cogit_dir

    # -- paths -------------------------------------------------------------

    def _ref_path(self, name: str) -> str:
        return os.path.join(self.cogit_dir, *name.split("/"))

    def _log_path(self, name: str) -> str:
        return os.path.join(self.cogit_dir, "logs", *name.split("/"))

    # -- HEAD ---------------------------------------------------------------

    def read_head_raw(self) -> str:
        path = os.path.join(self.cogit_dir, "HEAD")
        try:
            with open(path, "r", encoding="utf-8") as handle:
                return handle.read().strip()
        except FileNotFoundError as exc:
            raise CorruptionError("refs: HEAD missing") from exc

    def parse_head(self, content: str):
        """Return ('symbolic', refname) or ('detached', oid) for raw HEAD content."""
        if content.startswith("ref: "):
            refname = content[len("ref: ") :].strip()
            validate_ref_name(refname)
            return "symbolic", refname
        if is_oid(content):
            return "detached", content
        raise CorruptionError(f"refs: HEAD invalid: '{content}'")

    def read_head(self):
        return self.parse_head(self.read_head_raw())

    def write_head(self, value, old_head_target, actor, operation, reason, timestamp, expected_raw=None):
        """Point HEAD at 'ref: refs/heads/x' or a detached oid, with reflog.

        expected_raw, when given, is the raw HEAD content the caller observed;
        it is re-checked under the lock so a concurrent HEAD move fails with
        exit code 4 instead of being silently overwritten (COG-014).
        """
        if value.startswith("ref: "):
            validate_ref_name(value[len("ref: ") :])
        elif not is_oid(value):
            raise UserError(f"refs: invalid HEAD value '{value}'")
        path = os.path.join(self.cogit_dir, "HEAD")
        self._locked_replace(path, value + "\n", expected=expected_raw)
        self.append_reflog("HEAD", old_head_target, self._head_log_target(value), actor, operation, reason, timestamp)

    def _head_log_target(self, head_value: str) -> str:
        """Reflog records where HEAD logically points (oid when resolvable)."""
        if head_value.startswith("ref: "):
            target = self.read_ref(head_value[len("ref: ") :])
            return target if target else head_value[len("ref: ") :]
        return head_value

    # -- plain refs ----------------------------------------------------------

    def read_ref(self, name: str):
        validate_ref_name(name)
        path = self._ref_path(name)
        if not os.path.isfile(path):
            return None
        with open(path, "r", encoding="utf-8") as handle:
            content = handle.read().strip()
        if not is_oid(content):
            raise CorruptionError(f"refs: {name} has invalid target '{content}'")
        return content

    def list_refs(self, prefix: str):
        """Yield (refname, target) under a prefix like 'refs/heads'."""
        base = self._ref_path(prefix)
        if not os.path.isdir(base):
            return
        for dirpath, _dirnames, filenames in os.walk(base):
            for filename in sorted(filenames):
                if filename.endswith(".lock"):
                    continue
                full = os.path.join(dirpath, filename)
                rel = os.path.relpath(full, self._ref_path(""))
                refname = rel.replace(os.sep, "/")
                yield refname, self.read_ref(refname)

    def update_ref(self, name, new_target, expected_old, actor, operation, reason, timestamp):
        """Atomically move a ref with an old-target check, then append reflog."""
        validate_ref_name(name)
        if not is_oid(new_target):
            raise UserError(f"refs: invalid target '{new_target}'")
        path = self._ref_path(name)
        os.makedirs(os.path.dirname(path), exist_ok=True)
        lock_path = path + ".lock"
        try:
            fd = os.open(lock_path, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
        except FileExistsError as exc:
            raise ConcurrentUpdateError(f"refs: {name} is locked by another process") from exc
        try:
            current = self.read_ref(name)
            if current != expected_old:
                raise ConcurrentUpdateError(
                    f"refs: {name} moved (expected {expected_old or 'null'}, found {current or 'null'})"
                )
            with os.fdopen(fd, "w", encoding="utf-8") as handle:
                fd = None
                handle.write(new_target + "\n")
                handle.flush()
                os.fsync(handle.fileno())
            os.rename(lock_path, path)
        except Exception:
            if fd is not None:
                os.close(fd)
            if os.path.exists(lock_path):
                os.unlink(lock_path)
            raise
        self.append_reflog(name, expected_old, new_target, actor, operation, reason, timestamp)

    def delete_lockfile_safe(self, name):
        lock_path = self._ref_path(name) + ".lock"
        if os.path.exists(lock_path):
            os.unlink(lock_path)

    # -- reflog ---------------------------------------------------------------

    def append_reflog(self, name, old_target, new_target, actor, operation, reason, timestamp):
        if actor and any(ch.isspace() for ch in actor):
            raise UserError("reflog: actor must not contain whitespace")
        reason = " ".join(str(reason).splitlines()) or "-"
        line = f"{old_target or 'null'} {new_target} {timestamp} {actor} {operation}: {reason}\n"
        path = self._log_path(name)
        os.makedirs(os.path.dirname(path), exist_ok=True)
        try:
            with open(path, "a", encoding="utf-8") as handle:
                handle.write(line)
                handle.flush()
                os.fsync(handle.fileno())
        except OSError as exc:
            raise CorruptionError(
                f"reflog: {name} moved but journal append failed; operational history incomplete: {exc}"
            ) from exc

    def read_reflog(self, name):
        """Return parsed reflog entries, oldest first."""
        path = self._log_path(name)
        if not os.path.isfile(path):
            return []
        entries = []
        with open(path, "r", encoding="utf-8") as handle:
            for line_no, line in enumerate(handle, 1):
                line = line.rstrip("\n")
                if not line:
                    continue
                match = REFLOG_LINE_RE.match(line)
                if not match:
                    raise CorruptionError(f"reflog: {name}:{line_no} unparseable line")
                entries.append(match.groupdict())
        return entries

    # -- shared helpers ---------------------------------------------------------

    def _locked_replace(self, path, content, expected=None):
        """Replace a mutable file through <path>.lock + atomic rename.

        When expected is not None, the current content is re-read under the
        lock and must match (stripped) or ConcurrentUpdateError is raised.
        """
        lock_path = path + ".lock"
        try:
            fd = os.open(lock_path, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
        except FileExistsError as exc:
            raise ConcurrentUpdateError(f"refs: {path} is locked by another process") from exc
        try:
            if expected is not None:
                try:
                    with open(path, "r", encoding="utf-8") as current_handle:
                        current = current_handle.read().strip()
                except FileNotFoundError:
                    current = None
                if current != expected.strip():
                    raise ConcurrentUpdateError(
                        f"refs: {os.path.basename(path)} moved concurrently "
                        f"(expected '{expected.strip()}', found '{current}')"
                    )
            with os.fdopen(fd, "w", encoding="utf-8") as handle:
                fd = None
                handle.write(content)
                handle.flush()
                os.fsync(handle.fileno())
            os.rename(lock_path, path)
        except Exception:
            if fd is not None:
                os.close(fd)
            if os.path.exists(lock_path):
                os.unlink(lock_path)
            raise

File: crates/cogit-core/src/refs.rs
//! Refs, HEAD, and reflogs: lockfile updates with old-target checks.
//! Port of prototype/cogit/refs.py (ADR-0010 protocol, COG-014 HEAD checks).

use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

use serde_json::{json, Value};

use crate::error::{CoreError, Result};
use crate::objects::is_oid;

pub fn validate_ref_name(name: &str) -> Result<()> {
    let invalid = || CoreError::User(format!("refs: invalid ref name '{name}'"));
    if name.is_empty()
        || name.starts_with('/')
        || name.ends_with('/')
        || name.ends_with(".lock")
        || name.contains("@{")
        || name.contains('\\')
        || name.contains("..")
    {
        return Err(invalid());
    }
    for segment in name.split('/') {
        let seg_ok = !segment.is_empty()
            && segment
                .bytes()
                .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'.' || b == b'_' || b == b'-');
        if !seg_ok {
            return Err(CoreError::User(format!(
                "refs: invalid ref segment '{segment}' in '{name}'"
            )));
        }
    }
    Ok(())
}

#[derive(Debug, PartialEq, Eq)]
pub enum Head {
    Symbolic(String),
    Detached(String),
}

pub struct RefStore {
    cogit_dir: PathBuf,
}

impl RefStore {
    pub fn new(cogit_dir: &Path) -> Self {
        RefStore {
            cogit_dir: cogit_dir.to_path_buf(),
        }
    }

    fn ref_path(&self, name: &str) -> PathBuf {
        name.split('/').fold(self.cogit_dir.clone(), |p, s| p.join(s))
    }

    fn log_path(&self, name: &str) -> PathBuf {
        name.split('/').fold(self.cogit_dir.join("logs"), |p, s| p.join(s))
    }

    // -- HEAD -----------------------------------------------------------------

    pub fn read_head_raw(&self) -> Result<String> {
        fs::read_to_string(self.cogit_dir.join("HEAD"))
            .map(|s| s.trim().to_owned())
            .map_err(|_| CoreError::Corruption("refs: HEAD missing".into()))
    }

    pub fn parse_head(&self, content: &str) -> Result<Head> {
        if let Some(refname) = content.strip_prefix("ref: ") {
            let refname = refname.trim();
            validate_ref_name(refname)?;
            return Ok(Head::Symbolic(refname.to_owned()));
        }
        if is_oid(content) {
            return Ok(Head::Detached(content.to_owned()));
        }
        Err(CoreError::Corruption(format!("refs: HEAD invalid: '{content}'")))
    }

    pub fn read_head(&self) -> Result<Head> {
        self.parse_head(&self.read_head_raw()?)
    }

    #[allow(clippy::too_many_arguments)]
    pub fn write_head(
        &self,
        value: &str,
        old_target: Option<&str>,
        actor: &str,
        operation: &str,
        reason: &str,
        timestamp: &str,
        expected_raw: Option<&str>,
    ) -> Result<()> {
        if let Some(refname) = value.strip_prefix("ref: ") {
            validate_ref_name(refname)?;
        } else if !is_oid(value) {
            return Err(CoreError::User(format!("refs: invalid HEAD value '{value}'")));
        }
        let path = self.cogit_dir.join("HEAD");
        self.locked_replace(&path, &format!("{value}\n"), expected_raw)?;
        let log_target = if let Some(refname) = value.strip_prefix("ref: ") {
            self.read_ref(refname)?.unwrap_or_else(|| refname.to_owned())
        } else {
            value.to_owned()
        };
        self.append_reflog("HEAD", old_target, &log_target, actor, operation, reason, timestamp)
    }

    // -- plain refs -------------------------------------------------------------

    pub fn read_ref(&self, name: &str) -> Result<Option<String>> {
        validate_ref_name(name)?;
        let path = self.ref_path(name);
        if !path.is_file() {
            return Ok(None);
        }
        let content = fs::read_to_string(&path)?.trim().to_owned();
        if !is_oid(&content) {
            return Err(CoreError::Corruption(format!(
                "refs: {name} has invalid target '{content}'"
            )));
        }
        Ok(Some(content))
    }

    pub fn list_refs(&self, prefix: &str) -> Result<Vec<(String, String)>> {
        let base = self.ref_path(prefix);
        let mut out = Vec::new();
        if !base.is_dir() {
            return Ok(out);
        }
        let mut stack = vec![base];
        while let Some(dir) = stack.pop() {
            let mut entries: Vec<_> = fs::read_dir(&dir)?.filter_map(|e| e.ok()).collect();
            entries.sort_by_key(|e| e.file_name());
            for entry in entries {
                let path = entry.path();
                if path.is_dir() {
                    stack.push(path);
                } else if !path.to_string_lossy().ends_with(".lock") {
                    let rel = path
                        .strip_prefix(&self.cogit_dir)
                        .expect("under cogit dir")
                        .to_string_lossy()
                        .replace(std::path::MAIN_SEPARATOR, "/");
                    if let Some(target) = self.read_ref(&rel)? {
                        out.push((rel, target));
                    }
                }
            }
        }
        out.sort();
        Ok(out)
    }

    #[allow(clippy::too_many_arguments)]
    pub fn update_ref(
        &self,
        name: &str,
        new_target: &str,
        expected_old: Option<&str>,
        actor: &str,
        operation: &str,
        reason: &str,
        timestamp: &str,
    ) -> Result<()> {
        validate_ref_name(name)?;
        if !is_oid(new_target) {
            return Err(CoreError::User(format!("refs: invalid target '{new_target}'")));
        }
        let path = self.ref_path(name);
        fs::create_dir_all(path.parent().expect("ref parent"))?;
        let lock_path = path.with_extension("lock");
        let lock = fs::OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(&lock_path)
            .map_err(|_| CoreError::Concurrent(format!("refs: {name} is locked by another process")))?;
        let result = (|| -> Result<()> {
            let current = self.read_ref(name)?;
            if current.as_deref() != expected_old {
                return Err(CoreError::Concurrent(format!(
                    "refs: {name} moved (expected {}, found {})",
                    expected_old.unwrap_or("null"),
                    current.as_deref().unwrap_or("null"),
                )));
            }
            let mut lock = &lock;
            lock.write_all(format!("{new_target}\n").as_bytes())?;
            lock.sync_all()?;
            fs::rename(&lock_path, &path)?;
            Ok(())
        })();
        if result.is_err() {
            let _ = fs::remove_file(&lock_path);
            return result;
        }
        self.append_reflog(name, expected_old, new_target, actor, operation, reason, timestamp)
    }

    // -- reflog --------------------------------------------------------------------

    #[allow(clippy::too_many_arguments)]
    pub fn append_reflog(
        &self,
        name: &str,
        old_target: Option<&str>,
        new_target: &str,
        actor: &str,
        operation: &str,
        reason: &str,
        timestamp: &str,
    ) -> Result<()> {
        if actor.chars().any(char::is_whitespace) {
            return Err(CoreError::User("reflog: actor must not contain whitespace".into()));
        }
        let reason = reason.split_whitespace().collect::<Vec<_>>().join(" ");
        let reason = if reason.is_empty() { "-".to_owned() } else { reason };
        let line = format!(
            "{} {} {} {} {}: {}\n",
            old_target.unwrap_or("null"),
            new_target,
            timestamp,
            actor,
            operation,
            reason
        );
        let path = self.log_path(name);
        fs::create_dir_all(path.parent().expect("log parent"))?;
        let mut file = fs::OpenOptions::new().create(true).append(true).open(&path).map_err(|e| {
            CoreError::Corruption(format!(
                "reflog: {name} moved but journal append failed; operational history incomplete: {e}"
            ))
        })?;
        file.write_all(line.as_bytes()).and_then(|_| file.sync_all()).map_err(|e| {
            CoreError::Corruption(format!(
                "reflog: {name} moved but journal append failed; operational history incomplete: {e}"
            ))
        })
    }

    /// Parsed reflog entries, oldest first.
    pub fn read_reflog(&self, name: &str) -> Result<Vec<Value>> {
        let path = self.log_path(name);
        if !path.is_file() {
            return Ok(Vec::new());
        }
        let mut entries = Vec::new();
        for (line_no, line) in fs::read_to_string(&path)?.lines().enumerate() {
            if line.trim().is_empty() {
                continue;
            }
            let parts: Vec<&str> = line.splitn(5, ' ').collect();
            let parsed = if parts.len() == 5 {
                parts[4].split_once(": ")
            } else {
                None
            };
            match parsed {
                Some((op, reason)) if !op.contains(' ') => entries.push(json!({
                    "old": parts[0],
                    "new": parts[1],
                    "ts": parts[2],
                    "actor": parts[3],
                    "op": op,
                    "reason": reason,
                })),
                _ => {
                    return Err(CoreError::Corruption(format!(
                        "reflog: {name}:{} unparseable line",
                        line_no + 1
                    )))
                }
            }
        }
        Ok(entries)
    }

    pub fn list_reflogs(&self) -> Result<Vec<String>> {
        let logs_dir = self.cogit_dir.join("logs");
        let mut out = Vec::new();
        if !logs_dir.is_dir() {
            return Ok(out);
        }
        let mut stack = vec![logs_dir.clone()];
        while let Some(dir) = stack.pop() {
            for entry in fs::read_dir(&dir)?.filter_map(|e| e.ok()) {
                let path = entry.path();
                if path.is_dir() {
                    stack.push(path);
                } else {
                    let rel = path
                        .strip_prefix(&logs_dir)
                        .expect("under logs")
                        .to_string_lossy()
                        .replace(std::path::MAIN_SEPARATOR, "/");
                    out.push(rel);
                }
            }
        }
        out.sort();
        Ok(out)
    }

    /// Trim a reflog to its newest `keep` entries (COG-024). Returns (kept, dropped).
    pub fn expire_reflog(&self, name: &str, keep: usize, dry_run: bool) -> Result<(usize, usize)> {
        if keep < 1 {
            return Err(CoreError::User("reflog expire: --keep must be >= 1".into()));
        }
        let entries = self.read_reflog(name)?;
        let dropped = entries.len().saturating_sub(keep);
        if dropped == 0 || dry_run {
            return Ok((entries.len() - dropped, dropped));
        }
        let path = self.log_path(name);
        let content = fs::read_to_string(&path)?;
        let lines: Vec<&str> = content.lines().filter(|l| !l.trim().is_empty()).collect();
        let tail: String = lines[lines.len() - keep..]
            .iter()
            .map(|l| format!("{l}\n"))
            .collect();
        self.locked_replace(&path, &tail, None)?;
        Ok((keep, dropped))
    }

    // -- shared helpers ---------------------------------------------------------------

    pub fn locked_replace(&self, path: &Path, content: &str, expected: Option<&str>) -> Result<()> {
        let lock_path = PathBuf::from(format!("{}.lock", path.display()));
        let lock = fs::OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(&lock_path)
            .map_err(|_| {
                CoreError::Concurrent(format!("refs: {} is locked by another process", path.display()))
            })?;
        let result = (|| -> Result<()> {
            if let Some(expected) = expected {
                let current = fs::read_to_string(path).ok().map(|s| s.trim().to_owned());
                if current.as_deref() != Some(expected.trim()) {
                    return Err(CoreError::Concurrent(format!(
                        "refs: {} moved concurrently (expected '{}', found '{}')",
                        path.file_name().unwrap_or_default().to_string_lossy(),
                        expected.trim(),
                        current.as_deref().unwrap_or("null"),
                    )));
                }
            }
            let mut lock = &lock;
            lock.write_all(content.as_bytes())?;
            lock.sync_all()?;
            fs::rename(&lock_path, path)?;
            Ok(())
        })();
        if result.is_err() {
            let _ = fs::remove_file(&lock_path);
        }
        result
    }
}
